python/cudf/cudf/_lib/scalar.pyx

```diff
@@ -97,61 +86,61 @@ cdef class DeviceScalar:
             A NumPy dtype.
         """
         self._dtype = dtype if dtype.kind != 'U' else cudf.dtype('object')
-        self._set_value(value, self._dtype)
-
-    def _set_value(self, value, dtype):
-        # IMPORTANT: this should only ever be called from __init__
-        valid = not _is_null_host_scalar(value)
-
-        if isinstance(dtype, cudf.core.dtypes.DecimalDtype):
-            _set_decimal_from_scalar(
-                self.c_value, value, dtype, valid)
-        elif isinstance(dtype, cudf.ListDtype):
-            _set_list_from_pylist(
-                self.c_value, value, dtype, valid)
-        elif isinstance(dtype, cudf.StructDtype):
-            _set_struct_from_pydict(self.c_value, value, dtype, valid)
+
+        if cudf.utils.utils.is_na_like(value):
+            value = None
+        else:
+            # TODO: For now we always deepcopy the input value to avoid
+            # overwriting the input values when replacing nulls. Since it's
+            # just host values it's not that expensive, but we could consider
+            # alternatives.
```

Comment:
It seems like the most natural thing here is for `pa.scalar` to accept an `na_value` in its constructor. How do we handle this in the column case? Can you remind me what the consequences would be for just allowing this to fail at the `pa.scalar` call below?

Reply:
I'm not sure that I understand your question. Are you asking whether pyarrow understands pandas.NA? If so, the answer is no:

```
>>> pa.scalar(pd.NA)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/scalar.pxi", line 1120, in pyarrow.lib.scalar
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Could not convert <NA> with type NAType: did not recognize Python value type when inferring an Arrow data type
```
so we have to replace those NAs with normal Python None. The consequence of not doing this would be failing for any scalar that contained NA (or NaT), including nested scalars.

Reply:
We connected offline and chatted about this 👍